File: applications/qnet-mobile/android/app/src/main/java/com/qnetmobile/DilithiumModule.kt
package com.qnetmobile

import com.facebook.react.bridge.*
import android.util.Base64
import android.util.Log

/**
 * QNet Dilithium3 Native Module for React Native
 *
 * Uses the pqclean reference C implementation (same as server's pqcrypto-dilithium 0.5).
 * Provides byte-perfect compatibility with the server-side Dilithium3 verification.
 *
 * Signature size : 3309 bytes (FIPS 204 / pqclean dilithium3)
 * Public key size: 1952 bytes
 * Secret key size: 4032 bytes (stored as hex, never leaves device)
 *
 * Seed management: secret key is re-derived from the activation-code seed on every use.
 */
class DilithiumModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        const val NAME = "DilithiumModule"
        const val PUBLIC_KEY_SIZE  = 1952
        const val SECRET_KEY_SIZE  = 4032
        const val SIGNATURE_SIZE   = 3309   // pqclean / pqcrypto-dilithium 0.5

        init {
            System.loadLibrary("dilithium_native")
        }
    }

    // ---- Native declarations ----

    /** Returns pk (1952 bytes) || sk (4032 bytes) = 5984 bytes */
    private external fun nativeGenerateKeypair(seedStr: String): ByteArray?

    /** Returns 3309-byte detached signature */
    private external fun nativeSign(skBytes: ByteArray, msgBytes: ByteArray): ByteArray?

    /** Returns true if signature is valid */
    private external fun nativeVerify(pkBytes: ByteArray, sigBytes: ByteArray, msgBytes: ByteArray): Boolean

    /** Runs self-test, logs results to logcat and returns status string */
    private external fun nativeCompatTest(): String?

    override fun getName(): String = NAME

    // Run the native compat test on init — verifies the pqclean C code is working
    init {
        Thread {
            try {
                val result = nativeCompatTest()
                Log.e("DILITHIUM_COMPAT", "=== PQCLEAN NATIVE COMPAT TEST ===")
                Log.e("DILITHIUM_COMPAT", "Result: $result")
                Log.e("DILITHIUM_COMPAT", "SIG_SIZE=$SIGNATURE_SIZE (pqcrypto-dilithium 0.5 compatible)")
            } catch (e: Exception) {
                Log.e("DILITHIUM_COMPAT", "ERROR: ${e.message}")
            }
        }.start()
    }

    /**
     * Generate Dilithium3 keypair from deterministic seed.
     * Returns { publicKey: hex, secretKey: hex, publicKeySize, secretKeySize }
     * secretKey is the raw sk bytes in hex — re-derived from seed when needed.
     */
    @ReactMethod
    fun generateKeypairFromSeed(seed: String, promise: Promise) {
        try {
            val combined = nativeGenerateKeypair(seed)
                ?: throw RuntimeException("nativeGenerateKeypair returned null")

            if (combined.size != PUBLIC_KEY_SIZE + SECRET_KEY_SIZE) {
                throw RuntimeException("Unexpected keypair size: ${combined.size}")
            }

            val pk = combined.copyOfRange(0, PUBLIC_KEY_SIZE)
            val sk = combined.copyOfRange(PUBLIC_KEY_SIZE, PUBLIC_KEY_SIZE + SECRET_KEY_SIZE)

            val result = Arguments.createMap()
            result.putString("publicKey", bytesToHex(pk))
            // Store sk as hex — it can always be re-derived from seed
            // but storing it avoids re-running keygen on every sign call
            result.putString("secretKey", bytesToHex(sk))
            result.putInt("publicKeySize", pk.size)
            result.putInt("secretKeySize", sk.size)
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("DILITHIUM_KEYGEN_ERROR", "Failed to generate Dilithium3 keypair: ${e.message}", e)
        }
    }

    /**
     * Sign a message with Dilithium3.
     * secretKeySeed: hex-encoded 4032-byte secret key (from generateKeypairFromSeed).
     *   If it looks like a raw seed string (not hex / wrong length), re-derive the keypair.
     * Returns signature in backend-compatible format:
     *   "dilithium_sig_{nodeId}_{base64}"
     * where base64 encodes:
     *   [signed_msg_len(4 LE)] [signature || message] [pk_len(4 LE)] [public_key]
     */
    @ReactMethod
    fun sign(
        message: String,
        secretKeySeed: String,
        publicKeyHex: String,
        nodeId: String,
        promise: Promise
    ) {
        try {
            val messageBytes = message.toByteArray(Charsets.UTF_8)

            // Resolve sk bytes: if hex of correct length use directly, else re-derive
            val skBytes: ByteArray = resolveSecretKey(secretKeySeed)
            val pkBytes: ByteArray = hexToBytes(publicKeyHex)

            val sigBytes = nativeSign(skBytes, messageBytes)
                ?: throw RuntimeException("nativeSign returned null")

            if (sigBytes.size != SIGNATURE_SIZE) {
                throw RuntimeException("Unexpected sig size: ${sigBytes.size} (expected $SIGNATURE_SIZE)")
            }

            // Build binary payload: [4 LE len(sig||msg)] [sig||msg] [4 LE len(pk)] [pk]
            val signedMessage = sigBytes + messageBytes
            val binaryData = ByteArray(4 + signedMessage.size + 4 + pkBytes.size)
            var offset = 0
            putU32LE(binaryData, offset, signedMessage.size); offset += 4
            System.arraycopy(signedMessage, 0, binaryData, offset, signedMessage.size); offset += signedMessage.size
            putU32LE(binaryData, offset, pkBytes.size); offset += 4
            System.arraycopy(pkBytes, 0, binaryData, offset, pkBytes.size)

            val base64Sig = Base64.encodeToString(binaryData, Base64.NO_WRAP)
            val formattedSignature = "dilithium_sig_${nodeId}_${base64Sig}"

            val result = Arguments.createMap()
            result.putString("signature", formattedSignature)
            result.putInt("signatureSize", sigBytes.size)
            result.putInt("totalBinarySize", binaryData.size)
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("DILITHIUM_SIGN_ERROR", "Failed to sign with Dilithium3: ${e.message}", e)
        }
    }

    /**
     * Verify a Dilithium3 signature (local verification / testing).
     */
    @ReactMethod
    fun verify(
        message: String,
        signatureHex: String,
        publicKeyHex: String,
        promise: Promise
    ) {
        try {
            val pkBytes  = hexToBytes(publicKeyHex)
            val sigBytes = hexToBytes(signatureHex)
            val msgBytes = message.toByteArray(Charsets.UTF_8)
            val valid = nativeVerify(pkBytes, sigBytes, msgBytes)
            promise.resolve(valid)
        } catch (e: Exception) {
            promise.reject("DILITHIUM_VERIFY_ERROR", "Failed to verify: ${e.message}", e)
        }
    }

    /**
     * Run pqclean compatibility test (native).
     * Generates keypair, signs, verifies. Logs PK/SIG hex for cross-check with Rust.
     */
    @ReactMethod
    fun compatibilityTest(promise: Promise) {
        try {
            val result = nativeCompatTest()
                ?: throw RuntimeException("nativeCompatTest returned null")
            Log.e("DILITHIUM_COMPAT", "compatibilityTest result: $result")
            val map = Arguments.createMap()
            map.putString("result", result)
            map.putString("sigSize", SIGNATURE_SIZE.toString())
            map.putBoolean("isPqclean", true)
            promise.resolve(map)
        } catch (e: Exception) {
            promise.reject("COMPAT_TEST_ERROR", e.message, e)
        }
    }

    // ---- Private helpers ----

    /**
     * Resolve secret key bytes.
     * If secretKeySeed is hex of length 2*SECRET_KEY_SIZE (8064 chars) — decode directly.
     * Otherwise treat as a seed string and re-derive keypair.
     */
    private fun resolveSecretKey(secretKeySeed: String): ByteArray {
        if (secretKeySeed.length == SECRET_KEY_SIZE * 2 && secretKeySeed.all { it in '0'..'9' || it in 'a'..'f' || it in 'A'..'F' }) {
            return hexToBytes(secretKeySeed)
        }
        // Legacy path: re-derive from seed string
        val combined = nativeGenerateKeypair(secretKeySeed)
            ?: throw RuntimeException("Failed to re-derive keypair from seed")
        return combined.copyOfRange(PUBLIC_KEY_SIZE, PUBLIC_KEY_SIZE + SECRET_KEY_SIZE)
    }

    private fun putU32LE(buf: ByteArray, offset: Int, value: Int) {
        buf[offset]   = (value and 0xFF).toByte()
        buf[offset+1] = ((value shr 8)  and 0xFF).toByte()
        buf[offset+2] = ((value shr 16) and 0xFF).toByte()
        buf[offset+3] = ((value shr 24) and 0xFF).toByte()
    }

    private fun bytesToHex(bytes: ByteArray): String =
        bytes.joinToString("") { "%02x".format(it) }

    private fun hexToBytes(hex: String): ByteArray {
        val len = hex.length
        require(len % 2 == 0) { "Odd hex length: $len" }
        val data = ByteArray(len / 2)
        for (i in 0 until len step 2) {
            data[i / 2] = ((Character.digit(hex[i], 16) shl 4) +
                            Character.digit(hex[i + 1], 16)).toByte()
        }
        return data
    }
}

File: applications/qnet-mobile/ios/QNetMobile/DilithiumModule/DilithiumModule.m
/**
 * DilithiumModule.m
 * React Native native module for iOS — Dilithium3 (ML-DSA-65) post-quantum signatures.
 *
 * Mirrors Android DilithiumModule.kt. All methods, return shapes, and binary
 * formats are byte-identical so JS code (DilithiumCrypto.js) works unchanged.
 *
 * C sources included in the Xcode build target:
 *   DilithiumModule/dilithium3/*.c
 *   DilithiumModule/common/fips202.c
 *   DilithiumModule/randombytes_ios.c
 *
 * Key sizes (FIPS 204 / pqclean dilithium3):
 *   Public key : 1952 bytes
 *   Secret key : 4032 bytes
 *   Signature  : 3309 bytes
 */

#import "DilithiumModule.h"
#import <React/RCTLog.h>
#import <Foundation/Foundation.h>

/* PQClean C API */
#include "dilithium3/api.h"
#include "dilithium3/sign.h"
#include "common/fips202.h"
#include "randombytes_custom.h"

#include <string.h>
#include <stdlib.h>

#define DILITHIUM_PK_SIZE  PQCLEAN_DILITHIUM3_CLEAN_CRYPTO_PUBLICKEYBYTES  /* 1952 */
#define DILITHIUM_SK_SIZE  PQCLEAN_DILITHIUM3_CLEAN_CRYPTO_SECRETKEYBYTES  /* 4032 */
#define DILITHIUM_SIG_SIZE PQCLEAN_DILITHIUM3_CLEAN_CRYPTO_BYTES           /* 3309 */

/* ---- Hex helpers ---- */

static NSString *bytesToHex(const uint8_t *bytes, size_t len) {
    NSMutableString *hex = [NSMutableString stringWithCapacity:len * 2];
    for (size_t i = 0; i < len; i++) {
        [hex appendFormat:@"%02x", bytes[i]];
    }
    return hex;
}

static BOOL hexToBytes(NSString *hex, uint8_t *out, size_t expected_len) {
    if (hex.length != expected_len * 2) return NO;
    const char *str = hex.UTF8String;
    for (size_t i = 0; i < expected_len; i++) {
        char hi = str[2*i], lo = str[2*i+1];
        int h = (hi >= '0' && hi <= '9') ? hi-'0' :
                (hi >= 'a' && hi <= 'f') ? hi-'a'+10 :
                (hi >= 'A' && hi <= 'F') ? hi-'A'+10 : -1;
        int l = (lo >= '0' && lo <= '9') ? lo-'0' :
                (lo >= 'a' && lo <= 'f') ? lo-'a'+10 :
                (lo >= 'A' && lo <= 'F') ? lo-'A'+10 : -1;
        if (h < 0 || l < 0) return NO;
        out[i] = (uint8_t)((h << 4) | l);
    }
    return YES;
}

/** Derive deterministic 32-byte seed via SHAKE-256 (same as Android). */
static void deriveSeedFromString(const char *str, size_t len, uint8_t out[32]) {
    shake256(out, 32, (const uint8_t *)str, len);
}

/* ---- 4-byte little-endian write ---- */
static void writeU32LE(uint8_t *buf, uint32_t value) {
    buf[0] = (uint8_t)(value & 0xFF);
    buf[1] = (uint8_t)((value >> 8)  & 0xFF);
    buf[2] = (uint8_t)((value >> 16) & 0xFF);
    buf[3] = (uint8_t)((value >> 24) & 0xFF);
}

@implementation DilithiumModule

RCT_EXPORT_MODULE(DilithiumModule)

/** All methods run on a serial queue to keep single-threaded seed state safe. */
- (dispatch_queue_t)methodQueue {
    return dispatch_queue_create("com.qnetmobile.dilithium", DISPATCH_QUEUE_SERIAL);
}

/**
 * generateKeypairFromSeed(seed: string) → { publicKey, secretKey, publicKeySize, secretKeySize }
 *
 * Deterministically generates a Dilithium3 keypair from the given seed string.
 * Seed is hashed with SHAKE-256 to produce a 32-byte entropy input for keygen.
 */
RCT_EXPORT_METHOD(generateKeypairFromSeed:(NSString *)seed
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    const char *seedStr = seed.UTF8String;
    size_t seedLen = strlen(seedStr);

    uint8_t seed32[32];
    deriveSeedFromString(seedStr, seedLen, seed32);
    dilithium_set_keygen_seed(seed32);

    uint8_t pk[DILITHIUM_PK_SIZE];
    uint8_t sk[DILITHIUM_SK_SIZE];
    int ret = PQCLEAN_DILITHIUM3_CLEAN_crypto_sign_keypair(pk, sk);
    dilithium_clear_keygen_seed();

    if (ret != 0) {
        reject(@"DILITHIUM_KEYGEN_ERROR",
               [NSString stringWithFormat:@"nativeGenerateKeypair failed: %d", ret],
               nil);
        return;
    }

    resolve(@{
        @"publicKey":     bytesToHex(pk, DILITHIUM_PK_SIZE),
        @"secretKey":     bytesToHex(sk, DILITHIUM_SK_SIZE),
        @"publicKeySize": @(DILITHIUM_PK_SIZE),
        @"secretKeySize": @(DILITHIUM_SK_SIZE),
    });
}

/**
 * sign(message, secretKeySeed, publicKeyHex, nodeId) → { signature, signatureSize, totalBinarySize }
 *
 * secretKeySeed: 8064-char hex of raw secret key bytes (from generateKeypairFromSeed).
 *   If not valid hex of correct length, treats as seed string and re-derives keypair.
 *
 * Signature format (identical to Android):
 *   "dilithium_sig_{nodeId}_{base64([4LE:len(sig||msg)] [sig||msg] [4LE:len(pk)] [pk])}"
 */
RCT_EXPORT_METHOD(sign:(NSString *)message
                  secretKeySeed:(NSString *)secretKeySeed
                  publicKeyHex:(NSString *)publicKeyHex
                  nodeId:(NSString *)nodeId
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    /* Resolve secret key bytes */
    uint8_t sk[DILITHIUM_SK_SIZE];
    BOOL isRawHex = (secretKeySeed.length == DILITHIUM_SK_SIZE * 2) &&
                    hexToBytes(secretKeySeed, sk, DILITHIUM_SK_SIZE);

    if (!isRawHex) {
        /* Legacy / seed path: re-derive keypair from seed string */
        const char *seedStr = secretKeySeed.UTF8String;
        uint8_t seed32[32];
        deriveSeedFromString(seedStr, strlen(seedStr), seed32);
        dilithium_set_keygen_seed(seed32);

        uint8_t pk_tmp[DILITHIUM_PK_SIZE];
        if (PQCLEAN_DILITHIUM3_CLEAN_crypto_sign_keypair(pk_tmp, sk) != 0) {
            dilithium_clear_keygen_seed();
            reject(@"DILITHIUM_SIGN_ERROR", @"Failed to re-derive keypair from seed", nil);
            return;
        }
        dilithium_clear_keygen_seed();
    }

    /* Resolve public key bytes */
    uint8_t pk[DILITHIUM_PK_SIZE];
    if (!hexToBytes(publicKeyHex, pk, DILITHIUM_PK_SIZE)) {
        reject(@"DILITHIUM_SIGN_ERROR",
               [NSString stringWithFormat:@"Invalid public key hex (expected %d bytes)", DILITHIUM_PK_SIZE],
               nil);
        return;
    }

    /* Sign */
    const uint8_t *msgBytes = (const uint8_t *)message.UTF8String;
    size_t msgLen = strlen(message.UTF8String);

    uint8_t sig[DILITHIUM_SIG_SIZE];
    size_t  sigLen = 0;
    int ret = PQCLEAN_DILITHIUM3_CLEAN_crypto_sign_signature(
                  sig, &sigLen, msgBytes, msgLen, sk);

    /* Zero secret key immediately after use — prevent key material in stack residue */
    memset(sk, 0, DILITHIUM_SK_SIZE);

    if (ret != 0 || sigLen != DILITHIUM_SIG_SIZE) {
        reject(@"DILITHIUM_SIGN_ERROR",
               [NSString stringWithFormat:@"nativeSign failed: ret=%d sigLen=%zu", ret, sigLen],
               nil);
        return;
    }

    /* Build binary payload:
     *   [4 LE: len(sig||msg)] [sig||msg] [4 LE: len(pk)] [pk]
     * Identical to Android DilithiumModule.kt */
    size_t signedMsgLen = sigLen + msgLen;
    size_t totalLen     = 4 + signedMsgLen + 4 + DILITHIUM_PK_SIZE;
    uint8_t *buf = (uint8_t *)malloc(totalLen);
    if (!buf) {
        reject(@"DILITHIUM_SIGN_ERROR", @"Memory allocation failed", nil);
        return;
    }

    size_t offset = 0;
    writeU32LE(buf + offset, (uint32_t)signedMsgLen); offset += 4;
    memcpy(buf + offset, sig, sigLen);                offset += sigLen;
    memcpy(buf + offset, msgBytes, msgLen);            offset += msgLen;
    writeU32LE(buf + offset, (uint32_t)DILITHIUM_PK_SIZE); offset += 4;
    memcpy(buf + offset, pk, DILITHIUM_PK_SIZE);

    NSData *binaryData = [NSData dataWithBytes:buf length:totalLen];
    free(buf);

    NSString *base64Sig = [binaryData base64EncodedStringWithOptions:0];
    NSString *formattedSignature = [NSString stringWithFormat:@"dilithium_sig_%@_%@",
                                    nodeId, base64Sig];

    resolve(@{
        @"signature":        formattedSignature,
        @"signatureSize":    @(sigLen),
        @"totalBinarySize":  @(totalLen),
    });
}

/**
 * verify(message, signatureHex, publicKeyHex) → boolean
 *
 * signatureHex: hex-encoded raw 3309-byte signature (not the formatted string).
 */
RCT_EXPORT_METHOD(verify:(NSString *)message
                  signatureHex:(NSString *)signatureHex
                  publicKeyHex:(NSString *)publicKeyHex
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    size_t sigLen = signatureHex.length / 2;
    uint8_t *sig = (uint8_t *)malloc(sigLen);
    uint8_t pk[DILITHIUM_PK_SIZE];

    if (!sig) {
        reject(@"DILITHIUM_VERIFY_ERROR", @"Memory allocation failed", nil);
        return;
    }

    BOOL sigOk = hexToBytes(signatureHex, sig, sigLen);
    BOOL pkOk  = hexToBytes(publicKeyHex, pk, DILITHIUM_PK_SIZE);

    if (!sigOk || !pkOk) {
        free(sig);
        reject(@"DILITHIUM_VERIFY_ERROR", @"Invalid hex input for verify", nil);
        return;
    }

    const uint8_t *msgBytes = (const uint8_t *)message.UTF8String;
    size_t msgLen = strlen(message.UTF8String);

    int ret = PQCLEAN_DILITHIUM3_CLEAN_crypto_sign_verify(
                  sig, sigLen, msgBytes, msgLen, pk);
    free(sig);

    resolve(@(ret == 0));
}

/**
 * compatibilityTest() → { result, sigSize, isPqclean }
 *
 * Runs the same self-test as Android: fixed seed → keygen → sign → verify.
 * Logs PK/SIG chunks to NSLog for cross-checking with Rust compat test.
 */
RCT_EXPORT_METHOD(compatibilityTest:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    const char *testSeed = "QNET_COMPAT_TEST_SEED_v1";
    const char *testMsg  = "compatibility_test_message";
    size_t msgLen = strlen(testMsg);

    uint8_t seed32[32];
    deriveSeedFromString(testSeed, strlen(testSeed), seed32);
    dilithium_set_keygen_seed(seed32);

    uint8_t pk[DILITHIUM_PK_SIZE];
    uint8_t sk[DILITHIUM_SK_SIZE];
    PQCLEAN_DILITHIUM3_CLEAN_crypto_sign_keypair(pk, sk);
    dilithium_clear_keygen_seed();

    uint8_t sig[DILITHIUM_SIG_SIZE];
    size_t  sigLen = 0;
    PQCLEAN_DILITHIUM3_CLEAN_crypto_sign_signature(
        sig, &sigLen, (const uint8_t *)testMsg, msgLen, sk);

    int ok = PQCLEAN_DILITHIUM3_CLEAN_crypto_sign_verify(
                 sig, sigLen, (const uint8_t *)testMsg, msgLen, pk);

    NSLog(@"=== PQCLEAN COMPAT TEST (iOS) ===");
    NSLog(@"PK_LEN=%d SIG_LEN=%zu SELF_VERIFY=%@",
          DILITHIUM_PK_SIZE, sigLen, ok == 0 ? @"true" : @"false");

    /* Log PK in 1000-char hex chunks (matches Android logcat format) */
    NSString *pkHex = bytesToHex(pk, DILITHIUM_PK_SIZE);
    for (int i = 0; i * 1000 < (int)pkHex.length; i++) {
        NSInteger start = i * 1000;
        NSInteger end   = MIN(start + 1000, (NSInteger)pkHex.length);
        NSLog(@"PQCLEAN_PK[%d]%@", i,
              [pkHex substringWithRange:NSMakeRange(start, end - start)]);
    }

    /* Log SIG in 1000-char hex chunks */
    NSString *sigHex = bytesToHex(sig, sigLen);
    for (int i = 0; i * 1000 < (int)sigHex.length; i++) {
        NSInteger start = i * 1000;
        NSInteger end   = MIN(start + 1000, (NSInteger)sigHex.length);
        NSLog(@"PQCLEAN_SIG[%d]%@", i,
              [sigHex substringWithRange:NSMakeRange(start, end - start)]);
    }

    NSString *result = [NSString stringWithFormat:
        @"OK:PK_LEN=%d:SIG_LEN=%zu:SELF=%@",
        DILITHIUM_PK_SIZE, sigLen, ok == 0 ? @"OK" : @"FAIL"];

    resolve(@{
        @"result":    result,
        @"sigSize":   @(DILITHIUM_SIG_SIZE),
        @"isPqclean": @YES,
    });
}

@end
